lib/config/configitem.cpp

```diff
@@ -554,6 +556,10 @@ bool ConfigItem::CommitNewItems(const ActivationContext::Ptr& context, WorkQueue
 						try {
 							item->m_Object->OnAllConfigLoaded();
 							notified_items++;
+
+							if (!registerEarly) {
+								item->m_Object->Register();
+							}
```

Comment:
Discussed with Alex offline and he convinced me not to just rely on `load_after`, otherwise [such a stupid config](https://github.com/Icinga/icingaweb2-module-director/blob/428a49f4330c8863094e29582e106c3d8f9ef9f5/library/Director/Objects/IcingaService.php#L503-L510) won't function properly.

Reply:
Regarding that "stupid config":

```
    if (!host) {
        var host = get_host(host_name)
    }
```

I expect this to be rendered within a `Service` definition as it uses the `host_name` attribute. `Service` specifies `load_after Host;`, so why wouldn't this function when relying on `load_after`?

Reply:
See OP _What Icinga does with config_. `load_after` itself would function – Services will still be committed after Hosts. But during commit get_host() looks up _registered_ objects. So the host has to be registered immediately after commit, not after OnAllConfigLoaded.

Reply:
> so why wouldn't this function when relying on `load_after`?

If the objects were only registered after their `OnAllConfigLoaded()` was triggered, such *stupid config* would not work because they are evaluated while the object is being committed. However, a `load_after` dependency does not necessarily mean that a service is blocked from being committed before its host is **committed** and **OnAllConfigLoaded** is triggered, it just ensures that a service is committed after the host types, but before `Host#OnAllConfigLoaded` is triggered and a `Service#OnAllConfigLoaded` is triggered after the host equivalents.

Reply:
So if I understand correctly, relying on `load_after` wouldn't actually break `get_host()` in that example but rather require `OnAllConfigLoaded()` and `Register()` to be called incrementally after each type is loaded. And this would basically violate what the name `OnAllConfigLoaded()` advertises and most likely cause problems with that strange cyclic relation that exists between `Zone` and `Endpoint`.

Yes, this PR started with my suggestion/question whether it would be possible to avoid exposing partially initialized objects at all instead of checking for partially initialized objects in Icinga DB specifically to fix this bug. In the meantime, it was brought to light that there exists code that relies on seeing objects before they were fully initialized, so that suggestion won't work out. This PR tries to make it work nonetheless by registering different objects at different times during their initialization phase, but to be honest, I don't think that's better than the initial suggestion to skip objects still being initialized in the Icinga DB feature. Quite the opposite: this PR makes the config loading even more complex without achieving the goal I initially had in mind (which turned out to be impossible as other code relies on what I suggested to prevent).

Reply:
> Quite the opposite: this PR makes the config loading even more complex without achieving the goal I initially had in mind (which turned out to be impossible as other code relies on what I suggested to prevent).

Exactly! We either refactor the behaviour and let a `load_after` dependency block until the parent type is committed and its `OnAllConfigLoaded()` is triggered all at once, which is no easy task without breaking something, or just use a damn `nullptr` check where it causes problems now.

Reply:
> I don't think that's better than the initial suggestion to skip objects still being initialized in the Icinga DB feature.

**For now**, only the Icinga DB feature turned out to suffer from a yet incomplete Downtime. Instead of relying on that...

> just use a damn nullptr check where it causes problems now.

... and **working around** that "leak" of uninited objects, I suggest to just plug it via this PR.

I mean, once OnAllConfigLoaded() gets called initially, I (and the code) expect further object lookups to return only finished objects. That's what I fixed here.

Reply:
> I mean, once OnAllConfigLoaded() gets called initially, I (and the code) expect further object lookups to return only finished objects.

Doesn't Icinga 2 behave exactly as you expect it to behave even without this PR? Isn't the problem that the Icinga DB feature accesses objects too soon that are not yet fully initialised, i.e. before their `OnAllConfigLoaded` signal is triggered simply by going directly through the `ctype` objects vector? This PR just introduces an edge case and diverges runtime created and file based objects and makes the codebase even harder to follow what is going on. If you are keen on this, then please just fix this by using the existing `load_after` feature, which I doubt will make it into the upcoming bug fixe releases. By the way, there is also another PR from last week that addresses this issue #10151.